### PR TITLE
Fix crash in dev branch.

### PR DESCRIPTION
**Pull request by sipertruk**
From my PR https://github.com/ratkosrb/server/pull/12

### CODE REVIEW

File: src/game/Objects/Player.cpp [hunk 1]
```diff
@@ -1158,9 +1158,10 @@ struct UpdateAttackersCombatHelper
     explicit UpdateAttackersCombatHelper(Player* _player) : player(_player) {}
     void operator()(Unit* unit) const
     {
-        for (Unit* attacker : unit->getAttackers())
-            if (Creature* creature = attacker->ToCreature())
-                if (attacker->getVictim() == unit)
+        Unit::AttackerSet const attackers = unit->getAttackers();
+        for (Unit::AttackerSet::const_iterator itr = attackers.begin(); itr != attackers.end(); ++itr)
```
Comment (ryancheung) [inline src/game/Objects/Player.cpp:1162]: Can I ask why the simpler for-loop syntax cause a crash?

Comment (bladezLH) [reply src/game/Objects/Player.cpp:1162]: ^ all this is is a syntax change, i assume there is some concurrent modification going on

Comment (ryancheung) [reply src/game/Objects/Player.cpp:1162]: OK, I see

Comment (sipertruk) [reply src/game/Objects/Player.cpp:1162]: The loop is slightly different by the use of const_iterator but that's not the point.
That is copying the structure first like it's done elsewhere in the core, I couldn't reproduce the crash afterwards.

Comment (ratkosrb) [reply src/game/Objects/Player.cpp:1162]: Maybe getAttackers contains a nullptr so it crashes when you call ToCreature?

You can replace:
`if (Creature* creature = (*itr)->ToCreature())`

with

`if (Creature* creature = (*itr) && (*itr)->IsCreature() ? (*itr)->ToCreature() : nullptr)`

**Comment by sipertruk** [comment]
>#0  0x00007ffff667b863 in std::_Rb_tree_increment(std::_Rb_tree_node_base*) () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#1  0x0000555556175b27 in std::_Rb_tree_const_iterator<Unit*>::operator++ (this=0x7fffb5bfda40) at /usr/include/c++/7/bits/stl_tree.h:366
#2  0x0000555556296d5b in UpdateAttackersCombatHelper::operator() (this=0x7fffb5bfdad0, unit=0x7fffb7f653c0) at /home/sipertruk/git/core/src/game/Objects/Player.cpp:1161
#3  0x00005555562493e9 in Player::Update (this=0x7fffb7f653c0, update_diff=73, p_time=73) at /home/sipertruk/git/core/src/game/Objects/Player.cpp:1382

and
>#0  0x0000555555e8da62 in Object::GetTypeId (this=0x1) at /home/sipertruk/git/core/src/game/Objects/Object.h:343
#1  0x0000555555e8e096 in Object::IsCreature (this=0x1) at /home/sipertruk/git/core/src/game/Objects/Object.h:567
#2  0x0000555555fb31f0 in Object::ToCreature (this=0x1) at /home/sipertruk/git/core/src/game/Objects/Object.h:568
#3  0x0000555556296d21 in UpdateAttackersCombatHelper::operator() (this=0x7fffbdc91ad0, unit=0x7fffbfb005c0) at /home/sipertruk/git/core/src/game/Objects/Player.cpp:1162
#4  0x00005555562493e9 in Player::Update (this=0x7fffbfb005c0, update_diff=46, p_time=46) at /home/sipertruk/git/core/src/game/Objects/Player.cpp:1382

**Comment by bladezLH** [comment]
Yea, so it's due to modifying the set while iterating (SelectHostileTarget() can affect the unit's m_attackers). Creating a copy technically avoids this.

**Comment by bladezLH** [comment]
I'll be expanding upon this change a bit to resolve another crash related to instance switching shortly

**Pull request merged**